tfx/extensions/google_cloud_kubernetes/runner.py

```diff
@@ -0,0 +1,245 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Helper class to start TFX multi-worker training jobs on GKE."""
+
+import json
+from typing import Any, Dict, List, Text
+
+from absl import logging
+
+from tfx import types
+from tfx import version
+from tfx.types import artifact_utils
+from tfx.utils import kube_utils
+from kubernetes.client.rest import ApiException
+import kubernetes.client as client
+
+# Default TFX container image to use in GKE Training. For GPU training,
+# specify a custom image in executor.TRAINING_ARGS_KEY.
+_TFX_IMAGE = 'tensorflow/tfx:%s' % (version.__version__)
+
+_COMMAND = ["python", "-m", "tfx.scripts.run_executor"]
+
+
+def _build_pod_names(num_workers: int, unique_id: Text) -> List[Text]:
+  return ['training-worker-{}-{}'.format(unique_id,
+                                         i) for i in range(num_workers)]
+
+
+def _build_service_names(num_workers: int, unique_id: Text) -> List[Text]:
+  return ['training-service-{}-{}'.format(unique_id,
+                                          i) for i in range(num_workers)]
+
+
+def _pod_is_done(resp: client.V1Pod):
+  return kube_utils.PodPhase(resp.status.phase).is_done
+
+
+def create_worker_pods(job_args: List[Text],
+                       training_inputs: Dict[Text, Any],
+                       unique_id: Text):
+  """Create worker pods for multi-worker training."""
+  tfx_image = training_inputs.get('tfx_image', _TFX_IMAGE)
+  num_workers = training_inputs.get('num_workers', 1)
+  num_gpus_per_worker = training_inputs.get('num_gpus_per_worker', 0)
+
+  api_instance = kube_utils.make_core_v1_api()
+  service_names = _build_service_names(num_workers=num_workers,
+                                       unique_id=unique_id)
+  pod_names = _build_pod_names(num_workers=num_workers, unique_id=unique_id)
+  worker_hosts = ['{}:5000'.format(svc_name) for svc_name in service_names]
+
+  # TODO(ericlege): consider using a jinja2 template instead
+  for i in range(num_workers):
+    tf_config = json.dumps({
+        'cluster': {
+            'worker': worker_hosts
+        },
+        'task': {'type': 'worker', 'index': i}
+    })
+    pod = client.V1Pod(
+        metadata=client.V1ObjectMeta(
+            name=pod_names[i],
+            labels={
+                'name': 'training',
+                'id': unique_id,
+                'task': str(i),
+            },
+        ),
+        spec=client.V1PodSpec(
+            containers=[
+                client.V1Container(
+                    name='worker-pod',
+                    image=tfx_image,
+                    command=_COMMAND,
+                    args=job_args,
+                    security_context=client.V1SecurityContext(
+                        privileged=True,
```

Comment:
What happens when we don't use privileged?

Reply:
privileged=True grants the necessary container privilege for users' training script to use the TF profiler for profiling. Since many training scripts contains the profiler callback by default (i.e. cifar10 example), I think it would be most convenient to grant this by default.